In [4]:
import sys
sys.path.append('../')

from utils.utils import excute_cmd, execute_cmd_realtime

# To allow auto reload to this notebook after modifying any external file imported
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [10]:
command = 'python ../train.py --help'

# run the function to excute the command
excute_cmd(command, log=True)

usage: train.py [-h] [--train_path TRAIN_PATH] [--valid_path VALID_PATH]
                [--output OUTPUT] [--experiment_name EXPERIMENT_NAME]
                [--network_name NETWORK_NAME] [--max_epochs MAX_EPOCHS]
                [--batch_size BATCH_SIZE] [--base_lr BASE_LR]
                [--patience PATIENCE] [--img_size IMG_SIZE] [--seed SEED]
                [--verbose VERBOSE] [--normalize_attn]

options:
  -h, --help            show this help message and exit
  --train_path TRAIN_PATH
                        root dir for training data
  --valid_path VALID_PATH
                        root dir for validation data
  --output OUTPUT       output dir for saving results
  --experiment_name EXPERIMENT_NAME
                        experiment name
  --network_name NETWORK_NAME
                        network name
  --max_epochs MAX_EPOCHS
                        maximum epoch number to train
  --batch_size BATCH_SIZE
                        batch_size per gpu
  --base_lr BASE_LR     ne

We will be excuting all training commands here to keep track of all experiments. This notebook will only contain training logs.

In [4]:
command = 'python ../train.py --train_path "../datasets/challenge1/train" \
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "20" \
                        --batch_size "32" \
                        --verbose "1"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

2023-11-27 14:28:06.793 | INFO     | __main__:<module>:36 - Excuting training pipeline with 20 epochs...
2023-11-27 14:28:06.804 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/train
2023-11-27 14:28:06.869 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/val
2023-11-27 14:28:06.917 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for train split.
2023-11-27 14:28:06.917 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for val split.
2023-11-27 14:28:06.953 | INFO     | experiments.ClassifierExperiment:__init__:57 - Creating a classifier experiment using VGG16_BN_Attention network.
2023-11-27 14:28:15.096 | INFO     | networks.VGG16:__init__:107 - Using VGG16_BN_Attention with configurations: num_classes='2'
2023-11-27 14:28:17.769 | INFO     | experiments.ClassifierExperiment:run:233 - Experiment started.
2023-11-27 14:48:27.878 | INFO     | 

Experimenting the same network with same configuration but using `--normalize_attn` flag to normalize the network by using softmax instead of sigmoid for attention map 

In [9]:
command = 'python ../train.py --train_path "../datasets/challenge1/train" \
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --normalize_attn \
                        --max_epochs "20" \
                        --batch_size "32" \
                        --verbose "1"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

2023-11-27 20:34:37.132 | INFO     | __main__:<module>:38 - Excuting training pipeline with 20 epochs...
2023-11-27 20:34:37.143 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/train
2023-11-27 20:34:37.206 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/val
2023-11-27 20:34:37.224 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for train split.
2023-11-27 20:34:37.225 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for val split.
2023-11-27 20:34:37.260 | INFO     | experiments.ClassifierExperiment:__init__:57 - Creating a classifier experiment using VGG16_BN_Attention network.
2023-11-27 20:34:39.774 | INFO     | experiments.ClassifierExperiment:__init__:73 - Normalizing the attention network.
2023-11-27 20:34:39.775 | INFO     | networks.VGG16:__init__:107 - Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='T

We introduced model weights initialization as the official implementation using kaiming_normal, and re-run the first experiment on more epochs.

In [2]:
command = 'python ../train.py --train_path "../datasets/challenge1/train" \
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "30" \
                        --batch_size "32" \
                        --verbose "1"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

2023-11-28 17:07:41.238 | INFO     | __main__:<module>:38 - Excuting training pipeline with 30 epochs...
2023-11-28 17:07:41.263 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/train
2023-11-28 17:07:41.365 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/val
2023-11-28 17:07:41.388 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for train split.
2023-11-28 17:07:41.388 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for val split.
2023-11-28 17:07:41.780 | INFO     | experiments.ClassifierExperiment:__init__:57 - Creating a classifier experiment using VGG16_BN_Attention network.
2023-11-28 17:07:51.702 | INFO     | networks.VGG16:__init__:107 - Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
2023-11-28 17:07:59.650 | INFO     | experiments.ClassifierExperiment:run:238 - Experiment started.
2023-11-28 17

Experimenting `AdvancedHairAugmentation` in the transformation.

In [2]:
command = 'python ../train.py --train_path "../datasets/challenge1/train" \
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "30" \
                        --batch_size "32" \
                        --verbose "1"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

2023-11-30 17:44:20.598 | INFO     | __main__:<module>:38 - Excuting training pipeline with 30 epochs...
2023-11-30 17:44:20.629 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/train
2023-11-30 17:44:20.706 | INFO     | data_prep.dataset_loader:LoadData:24 - Loading the data from ../datasets/challenge1/val
2023-11-30 17:44:20.727 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for train split.
2023-11-30 17:44:20.727 | INFO     | data_prep.dataset:__init__:35 - Creating a Dataset instance for val split.
2023-11-30 17:44:20.791 | INFO     | experiments.ClassifierExperiment:__init__:65 - Creating a classifier experiment using VGG16_BN_Attention network.
2023-11-30 17:44:31.335 | INFO     | networks.VGG16:__init__:107 - Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
2023-11-30 17:44:34.680 | INFO     | experiments.ClassifierExperiment:run:352 - Experiment started.
2023-11-30 18

Training with 5 folds CV using StratifiedKFold and averaging all metrics at the end. Also, evaluate all 5 models using majority voting in the next cell.

In [4]:
command = 'python ../train_cv.py --train_path "../datasets/challenge1/train" \
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierExperimentCV" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "20" \
                        --num_folds "5" \
                        --batch_size "32" \
                        --verbose "1"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 20 epochs and 5 folds.
Loading the data from ../datasets/challenge1/train
Loading the data from ../datasets/challenge1/val
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Valid loss improved from inf to 0.386298. Saving model ...
[CV 1/5]: Epoch: 01/20 | epoch time: 19.0m 27.83s | lr: 1.00000e-04 | train/loss: 0.44440 | val/loss: 0.38630 | val/accuracy: 0.82653 | val/AUC: 0.82519 | val/Kappa: 0.65207
Valid loss improved from 0.386298 to 0.355428. Saving model ...
[CV 1/5]: Epoch: 02/20 | epoch time: 19.0m 1.74s | lr: 1.00000e-04 | train/loss: 0.38351 | val/loss: 0.35543 | val/accuracy: 0.83785 | val/AUC: 0.83676 | val/Kappa: 0.67493
Valid loss improved from 0.355428 to 0.346949. Saving model ...
[CV 1/5]: Epoch: 03/20 | epoch time: 17.0m 58.93s | lr: 1.00000e-04 | train/loss: 0.36290 | val/loss: 0.34695 | val/accuracy: 0.85154 | val/AUC: 0.85048 | val

Running inference on the previous experiment using the validation dataset.

In [3]:
command = 'python ../inference.py --test_path "../datasets/challenge1/val" \
                    --experiment_name "ClassifierExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "20" \
                    --batch_size "32" \
                    --timeframe "2023-12-03_2157" \
                    --verbose "2" \
                    --ensemble \
                    --report'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Constructed output path: outputs/ClassifierExperimentCV_224_epo20_bs32_lr0.0001_s42/2023-12-03_2157_VGG16_BN_Attention. Searching for models...
Found 5 models. Starting loading the models.
Loading data with 2 class labels from ../datasets/challenge1/val path...
Dataset labels: {'nevus': 0, 'others': 1} dictionary.
Loading the data from ../datasets/challenge1/val
Dataset length: 3796
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Namespace(test_path='../datasets/challenge1/val', output='outputs', experiment_name='ClassifierExperimentCV', network_name='VGG16_BN_Attention', max_epochs

Updated get_class_weight to use class get_class_weight. Using the training masks and attention masks during the training in `ClassifierSegExperiment`.

In [ ]:
command = 'python ../train.py --train_path "../datasets/challenge1/train" \
                        --train_masks_path "../datasets_masks/challenge1/train" \
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierSegExperiment" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "30" \
                        --batch_size "32" \
                        --verbose "2"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 30 epochs...
Loading data with 2 class labels...
Loading the data from ../datasets/challenge1/train
Loading the data from ../datasets/challenge1/val
Using segmentation masks for training...
train_images: 15195, train_masks: 15195, train_labels: 15195
Class weights: [0.98349515 1.01706827]
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using cross entropy loss.
Using dice loss.
Experiment started.
Epoch: 1 Train batch 10 loss: 0.6887211203575134, 2.1% complete
Epoch: 1 Train batch 20 loss: 0.4921400547027588, 4.2% complete
Epoch: 1 Train batch 30 loss: 0.4343116879463196, 6.3% complete


<p>
    Results was run on the server process and saved in the txt file. 
    
    2023-12-20 11:57:14.194 | INFO     | experiments.ClassifierSegExperiment:run_test:387 - Test Accuracy: 0.87460
    2023-12-20 11:57:14.194 | INFO     | experiments.ClassifierSegExperiment:run_test:388 - Test AUC: 0.87509
    2023-12-20 11:57:14.195 | INFO     | experiments.ClassifierSegExperiment:run_test:390 - Test Kappa: 0.74940
    2023-12-20 11:57:14.195 | INFO     | experiments.ClassifierSegExperiment:run_test:394 - Target 0 - Precision: 0.90039, Recall: 0.84723, F-score: 0.87300, Sensitivity: 0.8472294148109788, Specificity: 0.9029490616621983, Support: 1931
    2023-12-20 11:57:14.195 | INFO     | experiments.ClassifierSegExperiment:run_test:394 - Target 1 - Precision: 0.85093, Recall: 0.90295, F-score: 0.87617, Sensitivity: 0.9029490616621983, Specificity: 0.8472294148109788, Support: 1865
    
    2023-12-20 11:57:14.195 | INFO     | experiments.ClassifierSegExperiment:run_test:396 - 
    Classification Report:
                  precision    recall  f1-score   support
    
               0       0.90      0.85      0.87      1931
               1       0.85      0.90      0.88      1865
    
        accuracy                           0.87      3796
       macro avg       0.88      0.88      0.87      3796
    weighted avg       0.88      0.87      0.87      3796
    
</p>


In [16]:
command = 'python ../inference.py --test_path "../datasets/challenge1/val" \
                    --experiment_name "ClassifierSegExperiment" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "30" \
                    --batch_size "32" \
                    --timeframe "2023-12-19_2344" \
                    --verbose "2" \
                    --report'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Constructed output path: outputs/ClassifierSegExperiment_224_epo30_bs32_lr0.0001_s42/2023-12-19_2344_VGG16_BN_Attention. Searching for models...
Found 1 models. Starting loading the models.
Loading data with 2 class labels from ../datasets/challenge1/val path...
Dataset labels: {'nevus': 0, 'others': 1} dictionary.
Loading the data from ../datasets/challenge1/val
Dataset length: 3796
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Namespace(test_path='../datasets/challenge1/val', output='outputs', experiment_name='ClassifierSegExperiment', network_name='VGG16_BN_Attention', max_epochs=30, batch_size=32, base_lr=0.0001, img_size=224, seed=42, timeframe='2023-12-19_2344', verbose=2, multi=False, report=True, ensemble=False)

Inference: 100%|██████████| 119/119 [00:46<00:00,  2.59it/s]
Results exported to: outputs/ClassifierSegExperiment_224_epo30_bs32_lr0.0001_s42/2023-12-19_2344_VGG16_BN_Attention/ClassifierSegExperiment

Note, the center crop `A.CenterCrop`reduces the result significantlly! Was trained on few epochs and ended the training as it wasn't learning.

Results didn't improve with masks. Giving another try to build an ensemble and train it with the masks.

In [ ]:
command = 'python ../train_cv.py --train_path "../datasets/challenge1/train" \
                        --train_masks_path "../datasets_masks/challenge1/train" \
                        --valid_masks_path "../datasets_masks/challenge1/val"\
                        --valid_path "../datasets/challenge1/val" \
                        --experiment_name "ClassifierSegExperimentCV" \
                        --network_name "VGG16_BN_Attention" \
                        --max_epochs "30" \
                        --num_folds "5" \
                        --batch_size "32" \
                        --verbose "2"'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting training pipeline with 30 epochs and 5 folds.
Loading data with 2 class labels...
Loading the data from ../datasets/challenge1/train
Loading the data from ../datasets/challenge1/val
Using segmentation masks for training...
train_images: 15195, train_masks: 15195, train_labels: 15195
val_images: 3796, val_masks: 3796, val_labels: 3796
Class weights: [0.98342828 1.0171398 ]
Creating a classifier experiment using VGG16_BN_Attention network.
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using cross entropy loss.
Using dice loss.
[CV 1/5]: Epoch: 1 Train batch 10 loss: 0.6233866214752197, 2.1% complete
[CV 1/5]: Epoch: 1 Train batch 20 loss: 0.5773177742958069, 4.2% complete
[CV 1/5]: Epoch: 1 Train batch 30 loss: 0.5194941163063049, 6.3% complete
[CV 1/5]: Epoch: 1 Train batch 40 loss: 0.4522211253643036, 8.4% complete
[CV 1/5]: Epoch: 1 Train batch 50 loss: 0.45502814650535583, 10.5% complete
[CV 1/5]: Epoch: 1 Train batch 60 loss: 0.374016

Training on the above ensemble finished on the server. Evaluating below.

In [18]:
command = 'python ../inference.py --test_path "../datasets/challenge1/val" \
                    --experiment_name "ClassifierSegExperimentCV" \
                    --network_name "VGG16_BN_Attention" \
                    --max_epochs "30" \
                    --batch_size "32" \
                    --timeframe "2023-12-20_1507" \
                    --verbose "2" \
                    --ensemble \
                    --report'

# run the function to excute the command
_ = execute_cmd_realtime(command, log=True)

Excuting inference script...
Constructed output path: outputs/ClassifierSegExperimentCV_224_epo30_bs32_lr0.0001_s42/2023-12-20_1507_VGG16_BN_Attention. Searching for models...
Found 5 models. Starting loading the models.
Loading data with 2 class labels from ../datasets/challenge1/val path...
Dataset labels: {'nevus': 0, 'others': 1} dictionary.
Loading the data from ../datasets/challenge1/val
Dataset length: 3796
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Using VGG16_BN_Attention with configurations: num_classes='2', normalize_attn='False'
Namespace(test_path='../datasets/challenge1/val', output='outputs', experiment_name='ClassifierSegExperimentCV', network_name='VGG16_BN_Attention', max_

Using the mask improved the ensemble in some metrics a little bit.

In [ ]:
# Using `VGG16_BN_Attention_P4` with one attention block from pool layer 4.Added `A.CenterCrop` to all transforms (including the val/test)